pkg/plugin/processor/builtin/js/processor.go

```diff
@@ -0,0 +1,389 @@
+was an error processing the input record. Details about the error can be provided through a
+string argument in the constructor, e.g. ErrorRecord("details about the error").
+* RawData(): creates a raw data object. It's analogous to opencdc.RawData. Optionally, it
+accepts a string argument, which will be cast into a byte array, for example: record.Key = RawData("new key").
+* StructuredData(): creates a structured data (map-like) object.
+
+To find out what's possible with the JS processors, also refer to the documentation for 
+[goja](https://github.com/dop251/goja), which is the JavaScript engine we use.`,
+		Version:    "v0.1.0",
+		Author:     "Meroxa, Inc.",
+		Parameters: processorConfig{}.Parameters(),
+	}, nil
+}
+
+func (p *processor) Configure(ctx context.Context, m map[string]string) error {
+	cfg := processorConfig{}
+	err := sdk.ParseConfig(ctx, m, &cfg, cfg.Parameters())
+	if err != nil {
+		return cerrors.Errorf("failed parsing configuration: %w", err)
+	}
+
+	switch {
+	case cfg.Script != "" && cfg.ScriptPath != "":
+		return cerrors.New("only one of: [script, script.path] should be provided")
+	case cfg.Script != "":
+		p.src = cfg.Script
+	case cfg.ScriptPath != "":
+		file, err := os.ReadFile(cfg.ScriptPath)
+		if err != nil {
+			return cerrors.Errorf("error reading script from path %v: %w", cfg.ScriptPath, err)
+		}
+		p.src = string(file)
+	default:
+		return cerrors.New("one of: [script, script.path] needs to be provided")
+	}
+
+	return nil
+}
+
+func (p *processor) Open(context.Context) error {
+	runtime, err := p.newRuntime(p.logger)
+	if err != nil {
+		return cerrors.Errorf("failed initializing JS runtime: %w", err)
+	}
+
+	_, err = p.newFunction(runtime, p.src)
+	if err != nil {
+		return cerrors.Errorf("failed initializing JS function: %w", err)
+	}
+
+	p.gojaPool.New = func() any {
+		// create a new runtime for the function, so it's executed in a separate goja context
+		rt, _ := p.newRuntime(p.logger)
+		f, _ := p.newFunction(rt, p.src)
+		return &gojaContext{
+			runtime:  rt,
+			function: f,
+		}
+	}
+
+	return nil
+}
+
+func (p *processor) Process(_ context.Context, records []opencdc.Record) []sdk.ProcessedRecord {
+	g := p.gojaPool.Get().(*gojaContext)
+	defer p.gojaPool.Put(g)
+
+	jsRecs := p.toJSRecords(g.runtime, records)
+
+	result, err := g.function(goja.Undefined(), jsRecs)
+	if err != nil {
+		return p.makeProcessedRecords(sdk.ErrorRecord{Error: err}, len(records))
+	}
+
+	return p.toSDKRecords(g.runtime, result, len(records))
+}
+
+func (p *processor) Teardown(context.Context) error {
+	return nil
+}
+
+func (p *processor) newRuntime(logger log.CtxLogger) (*goja.Runtime, error) {
+	rt := goja.New()
+	require.NewRegistry().Enable(rt)
+
+	runtimeHelpers := map[string]interface{}{
+		"logger":         &logger,
+		"SingleRecord":   p.newSingleRecord(rt),
+		"FilterRecord":   p.newFilterRecord(rt),
+		"ErrorRecord":    p.newErrorRecord(rt),
```

Comment:
I'm not a fan of exposing this to the javascript processor, it needlessly complicates writing a JS processor. I'd rather keep the functionality as it was before, meaning that the processor gets 1 record at a time (we can loop through the batch and supply the records one by one), and if it returns `nil` it's considered a filtered record, if it returns a record we use it, anything else is considered an error.

Reply:
I'm aware that this means we'll call the processor multiple times when processing a batch and lose some performance, but if the user wants performance they wouldn't use the JS processor in the first place, it's already slow. And since we now have WASM as an alternative, I think we don't have to worry about the performance here and rather keep it simple.

Reply:
I was considering that too, but I voted for this one because it keeps the API very similar across languages.

Reply:
> Do you mean the API compared to WASM processors?

Yup!



> And once we have, that approach will require 4 functions to be implemented instead of 1

Can you elaborate a bit, why 4 functions?






> In any case, these processors will not be interchangeable

That's true, the code won't be interchangeable, but the idea is that the concepts should be. In other words, if you wrote a processor in WASM, ideally, you'll just skim through the JS docs and write a processor in JS.

Reply:
Well you have `Open`, `Configure`, `Read` and `Teardown`. Also you'll need to provide specifications, so there's probably 5. And that's where I'd offer the full functionality. Maybe there will be a simple way to write the WASM processor, similar to how the Go SDK provides `NewProcessorFunc`, and that could be the API that is similar to the `js` processor. But I see no reason to provide a complexer API here, for the reasons stated in my first answer.

Reply:
Gotcha. Well, I wouldn't say it's that much more complex It's just making the `Process()` function more similar to how it looks like across other languages. I don't mind changing it though, especially given the timeline.:)

Reply:
Done!